rfq/order.go

```diff
@@ -0,0 +1,320 @@
+	// which the remit applies.
+	Scid SerialisedScid
+
+	// AssetAmount is the amount of the tap asset that is being requested.
+	AssetAmount uint64
+
+	// MinimumChannelPayment is the minimum number of millisatoshis that
+	// must be sent in the HTLC.
+	MinimumChannelPayment lnwire.MilliSatoshi
+
+	// Expiry is the asking price expiryDelay lifetime unix timestamp.
+	Expiry uint64
+}
+
+// NewChannelRemit creates a new channel remit.
+func NewChannelRemit(quoteAccept rfqmsg.Accept) *ChannelRemit {
+	// Compute the serialised short channel ID (SCID) for the channel.
+	scid := SerialisedScid(quoteAccept.ShortChannelId())
+
+	return &ChannelRemit{
+		Scid:                  scid,
+		AssetAmount:           quoteAccept.AssetAmount,
+		MinimumChannelPayment: quoteAccept.AskPrice,
+		Expiry:                quoteAccept.Expiry,
+	}
+}
+
+// CheckHtlcCompliance returns an error if the given HTLC intercept descriptor
+// does not satisfy the subject channel remit.
+func (c *ChannelRemit) CheckHtlcCompliance(
+	htlc lndclient.InterceptedHtlc) error {
+
+	// Check that the channel SCID is as expected.
+	htlcScid := SerialisedScid(htlc.OutgoingChannelID.ToUint64())
+	if htlcScid != c.Scid {
+		return fmt.Errorf("htlc outgoing channel ID does not match "+
+			"remit's SCID (htlc_scid=%d, remit_scid=%d)", htlcScid,
+			c.Scid)
+	}
+
+	// Check that the HTLC amount is at least the minimum acceptable amount.
+	if htlc.AmountOutMsat < c.MinimumChannelPayment {
+		return fmt.Errorf("htlc out amount is less than the remit's "+
+			"minimum (htlc_out_msat=%d, remit_min_msat=%d)",
+			htlc.AmountOutMsat, c.MinimumChannelPayment)
+	}
+
+	// Lastly, check to ensure that the channel remit has not expired.
+	if time.Now().Unix() > int64(c.Expiry) {
+		return fmt.Errorf("channel remit has expired "+
+			"(expiry_unix_ts=%d)", c.Expiry)
+	}
+
+	return nil
+}
+
+// OrderHandlerCfg is a struct that holds the configuration parameters for the
+// order handler service.
+type OrderHandlerCfg struct {
+	// CleanupInterval is the interval at which the order handler cleans up
+	// expired accepted quotes from its local cache.
+	CleanupInterval time.Duration
+
+	// HtlcInterceptor is the HTLC interceptor. This component is used to
+	// intercept and accept/reject HTLCs.
+	HtlcInterceptor HtlcInterceptor
+
+	// AcceptHtlcEvents is a channel that receives accepted HTLCs.
+	AcceptHtlcEvents chan<- *AcceptHtlcEvent
+}
+
+// OrderHandler orchestrates management of accepted quote bundles. It monitors
+// HTLCs (Hash Time Locked Contracts), and determines acceptance/rejection based
+// on the terms of the associated accepted quote.
+type OrderHandler struct {
+	startOnce sync.Once
+	stopOnce  sync.Once
+
+	// cfg holds the configuration parameters for the RFQ order handler.
+	cfg OrderHandlerCfg
+
+	// channelRemits is a map of serialised short channel IDs (SCIDs) to
+	// associated active channel remits.
+	channelRemits map[SerialisedScid]*ChannelRemit
```

Comment:
nit: could probably use `lnutils.SyncMap` in a few of the new subsystems instead of a manual map + mtx.